meta/SearchResult.php

```diff
@@ -0,0 +1,171 @@
+<?php
+
+namespace dokuwiki\plugin\struct\meta;
+
+/**
+ * Class SearchResult
+ *
+ * Search is executed only once per request.
+ */
+class SearchResult
+{
+    /** @var Value[][] */
+    protected $rows = [];
+    /** @var array */
+    protected $pids = [];
+    protected $rids = [];
+    /** @var array */
+    protected $revs = [];
+    /** @var int */
+    protected $count = -1;
+
+    /**
+     * Construct SearchResult
+     *
+     * @param \PDOStatemen $res
+     * @param int $rangeBegin
+     * @param int $rangeEnd
+     * @param array $columns
+     * @param bool $pageidAndRevOnly
+     */
+    public function __construct($res, $rangeBegin, $rangeEnd, $columns, $pageidAndRevOnly)
+    {
+        while ($row = $res->fetch(\PDO::FETCH_ASSOC)) {
+            $this->increaseCount();
+            if ($this->getCount() < $rangeBegin) continue;
+            if ($rangeEnd && $this->getCount() >= $rangeEnd) continue;
+
+            $C = 0;
+            $resrow = [];
+            $isempty = true;
+            foreach ($columns as $col) {
+                $val = $row["C$C"];
+                if ($col->isMulti()) {
+                    $val = explode(Search::CONCAT_SEPARATOR, $val);
+                }
+                $value = new Value($col, $val);
+                $isempty &= $this->isEmptyValue($value);
+                $resrow[] = $value;
+                $C++;
+            }
+
+            // skip empty rows
+            if ($isempty && !$pageidAndRevOnly) {
+                $this->decreaseCount();
+                continue;
+            }
+
+            $this->addPid($row['PID']);
+            $this->addRid($row['rid']);
+            $this->addRev($row['rev']);
+            $this->addRow($resrow);
+        }
+
+        $res->closeCursor();
```

Comment:
I wonder if the cursor should be closed here or by the one who passes it... If this method does it, we should mention it in the method comment